### PR TITLE
Polynomial-based hashes

### PR DESCRIPTION
**Pull request by patrick-schultz**
* Implement arithemetic in the finite field of order 2^32, as
  polynomials over F_2 modulo the irreducible polynomial
  x^32 + x^7 + x^3 + x^2 + 1.

* Define class `PolyHash`, implementing polynomials over above field,
  using Horner's rule for polynomial evaluation.

* Use random polynomials to fill tables in tabulation hashes, for
  guaranteed 32-independent randomness.

### CODE REVIEW

File: src/test/scala/is/hail/utils/HashMethodsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package is.hail.utils
+
+import is.hail.SparkSuite
+import org.testng.annotations.Test
+
+class HashMethodsSuite extends SparkSuite {
+  @Test def testMultGF(): Unit = {
+    val ph = new PolyHash
```
Comment (jbloom22) [inline src/test/scala/is/hail/utils/HashMethodsSuite.scala]: you left off the parameters

File: src/test/scala/is/hail/utils/HashMethodsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    import ph._
+
+    val x1 = 3705006673L.toInt
+    val y1 = 2551778209L.toInt
+    assert(multGF(x1, y1) == 2272553040L.toInt)
```
Comment (jbloom22) [inline src/test/scala/is/hail/utils/HashMethodsSuite.scala]: i'm curious, what did you use to generate the right hand side?

Comment (patrick-schultz) [reply src/test/scala/is/hail/utils/HashMethodsSuite.scala]: I used Mathematica to test my implementation, then just hard coded a couple of random cases.

File: src/main/scala/is/hail/utils/HashMethods.scala [hunk 3]
```diff
@@ -121,3 +126,59 @@ class FiveIndepTabulationHash32(keyTable: Array[Long], derivedKeyTable: Array[In
...
+
+  // Can be done by the PCLMULQDQ "carryless multiply" instruction on x86 processors post ~2010.
+  // This would give a significant speed boost. Any way to do this from JVM?
+  def polyMult(a: Int, b: Int): Long = {
+    var result: Long = 0
+    if (b != 0) {
+      var aBuf: Int = a
+      var bBuf: Long = b & 0xffffffffL
+      while (aBuf != 0) {
+        if ((aBuf & 1) == 1) result ^= bBuf
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply src/main/scala/is/hail/utils/HashMethods.scala]: Good call. Or maybe to save the (relatively expensive) multiplication, I could try `result ^= bBuf & a(aBuf & 1)`, where `a = Array[Long](0, -1)`.

Do you think I should benchmark this now, or can I save it for a future PR? (Today is my last day before leaving for Thanksgiving, and I have several things to get done before I leave.)

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply src/main/scala/is/hail/utils/HashMethods.scala]: I went ahead and did some tests. Yours was the winner, over twice as fast as the branching version. The array one was twice as slow, and I understand why now. I also tried `result ^= bBuf & (0L - (aBuf & 1))`, which was barely slower than yours, which surprised me. I thought an & and a subtraction would be faster than a multiplication.

**Pull request merged**